drivers/interrupt_controller/gic.c

```diff
@@ -0,0 +1,230 @@
+#define	GICD_ICFGRn				(CONFIG_GIC_DIST_BASE + 0x0C00)
+#define	GICD_SGIR				(CONFIG_GIC_DIST_BASE + 0x0F00)
+
+#define GICC_CTRL				(CONFIG_GIC_CPU_BASE + 0x0000)
+#define GICC_PMR				(CONFIG_GIC_CPU_BASE + 0x0004)
+#define GICC_BPR				(CONFIG_GIC_CPU_BASE + 0x0008)
+#define GICC_IAR				(CONFIG_GIC_CPU_BASE + 0x000C)
+#define GICC_EOIR				(CONFIG_GIC_CPU_BASE + 0x0010)
+
+#define GICC_ENABLE	3
+#define GICC_DIS_BYPASS_MASK		0x1e0
+
+#define	NO_GIC_INT_PENDING	1023
+
+/* CPU Interrupt numbers */
+#define GIC_INT_GTMR			27
+#define	GIC_INT_FIQ				28
+#define	GIC_INT_TMR				29
+#define	GIC_INT_WD				30
+#define	GIC_INT_IRQ				31
+
+#define IRQ_TYPE_NONE			0x00000000
+#define IRQ_TYPE_EDGE_RISING	0x00000001
+#define IRQ_TYPE_EDGE_FALLING	0x00000002
+#define IRQ_TYPE_EDGE_BOTH		(IRQ_TYPE_EDGE_FALLING | \
+							IRQ_TYPE_EDGE_RISING)
+#define IRQ_TYPE_LEVEL_HIGH		0x00000004
+#define IRQ_TYPE_LEVEL_LOW		0x00000008
+#define IRQ_TYPE_LEVEL_MASK		(IRQ_TYPE_LEVEL_LOW | \
+							IRQ_TYPE_LEVEL_HIGH)
+#define IRQ_TYPE_SENSE_MASK		0x0000000f
+#define IRQ_TYPE_DEFAULT		IRQ_TYPE_SENSE_MASK
+
+#define	GIC_SPI_INT_BASE	32
+
+static void gic_dist_init(void)
+{
+	unsigned int gic_irqs, i;
+
+	gic_irqs = sys_read32(GICD_TYPER) & 0x1f;
+	gic_irqs = (gic_irqs + 1) * 32;
+	if (gic_irqs > 1020)
+		gic_irqs = 1020;
+
+	/*
+	 * Disable the forwarding of pending interrupts
+	 * from the Distributor to the CPU interfaces
+	 */
+	sys_write32(0, GICD_CTRL);
+
+	/*
+	 * Set all global interrupts to this CPU only.
+	 */
+	for (i = GIC_SPI_INT_BASE; i < gic_irqs; i += 4)
+		sys_write32(0x01010101, GICD_ITARGETSRn + i);
+
+	/*
+	 * Set all global interrupts to be level triggered, active low.
+	 */
+	for (i = GIC_SPI_INT_BASE; i < gic_irqs; i += 16)
+		sys_write32(0, GICD_ICFGRn + i / 4);
+
+	/*  Set priority on all global interrupts.   */
+	for (i = GIC_SPI_INT_BASE; i < gic_irqs; i += 4)
+		sys_write32(0xa0a0a0a0, GICD_IPRIORITYRn + i);
+
+	/* Set all interrupts to group 0 */
+	for (i = GIC_SPI_INT_BASE; i < gic_irqs; i += 32)
+		sys_write32(0, GICD_IGROUPRn + i / 8);
+
+	/*
+	 * Disable all interrupts.  Leave the PPI and SGIs alone
+	 * as these enables are banked registers.
+	 */
+	for (i = GIC_SPI_INT_BASE; i < gic_irqs; i += 32) {
+		sys_write32(0xffffffff, GICD_ICACTIVERn + i / 8);
+		sys_write32(0xffffffff, GICD_ICENABLERn + i / 8);
+	}
+
+	/*
+	 * Enable the forwarding of pending interrupts
+	 * from the Distributor to the CPU interfaces
+	 */
+	sys_write32(1, GICD_CTRL);
+}
+
+static void gic_cpu_init(void)
+{
+	int i;
+	u32_t val;
+
+	/*
+	 * Deal with the banked PPI and SGI interrupts - disable all
+	 * PPI interrupts, ensure all SGI interrupts are enabled.
+	 */
+	sys_write32(0xffffffff, GICD_ICACTIVERn);
+	sys_write32(0xffff0000, GICD_ICENABLERn);
+	sys_write32(0x0000ffff, GICD_ISENABLERn);
+
+	/*
+	 * Set priority on PPI and SGI interrupts
+	 */
+	for (i = 0; i < 32; i += 4)
+		sys_write32(0xa0a0a0a0, GICD_IPRIORITYRn + i);
```

Comment:
I see two possible paths to get the flags and priority from gic.c.  1) add an intr_set_priority() callback to the irq_next_level_api struct and call it from arch/arm/core/irq_manage.c:_irq_priority_set() OR 2) redefine _ARCH_IRQ_CONNECT for Cortex-R to pass flags_p instead of 0 to _ISR_DECLARE and get to the flags through the isr_sw_table.

Do you have a preference between these two, or another option?